contrib/parseq-restli-client/src/main/java/com/linkedin/restli/client/ParSeqRestClient.java

```diff
@@ -181,26 +195,72 @@ static String generateTaskName(final Request<?> request) {
     } else {
-      return task;
+      return task.withTimeout(timeout.getValue(), TimeUnit.MILLISECONDS);
     }
   }
 
+  private <T> Task<Response<T>> withD2Timeout(final Task<Response<T>> task, String timeoutTaskName) {
+    return task.transform(timeoutTaskName, (Try<Response<T>> tryGet) -> {
+      if (tryGet.isFailed() && tryGet.getError() instanceof TimeoutException) {
+        String timeoutExceptionMessage = "task: '" + task.getName() + "' " + timeoutTaskName;
+        return Failure.of(Exceptions.timeoutException(timeoutExceptionMessage));
+      } else {
+        return tryGet;
+      }
+    });
+  }
+
   private <T> Task<Response<T>> createTask(final String name, final Request<T> request,
       final RequestContext requestContext, RequestConfig config) {
     LOGGER.debug("createTask, name: '{}', config: {}", name, config);
+    return createTaskWithTimeout(name, request, requestContext, config);
+  }
+
+  // reconcile timeout specified in requestContext (if there is) with the given timeout to pick a tighter one
+  private int reconcileRequestTimeout(final RequestContext requestContext, Long timeout) {
+    int d2Timeout = timeout.intValue();
+    Object requestTimeout = requestContext.getLocalAttr(R2Constants.REQUEST_TIMEOUT);
+    if (requestTimeout instanceof Number) {
+      d2Timeout = Math.min(((Number)requestTimeout).intValue(), d2Timeout);
+    } else if (requestTimeout != null) {
+      LOGGER.error("Ignore invalid value for REQUEST_TIMEOUT set in request context: " + requestTimeout);
+    }
+    requestContext.putLocalAttr(R2Constants.REQUEST_TIMEOUT, d2Timeout);
+    return d2Timeout;
+  }
+
+  private <T> Task<Response<T>> createTaskWithTimeout(final String name, final Request<T> request,
+      final RequestContext requestContext, RequestConfig config) {
+    ConfigValue<Long> timeout = config.getTimeoutMs();
+    int reconciledTimeout = 0;
+    if (timeout.getValue() != null && timeout.getValue() > 0 && _d2RequestTimeoutEnabled) {
+      // reconcile timeout if request has already set timeout in its request context
+      reconciledTimeout = reconcileRequestTimeout(requestContext, timeout.getValue());
+    }
+    Task<Response<T>> requestTask;
     if (RequestGroup.isBatchable(request, config)) {
-      return withTimeout(createBatchableTask(name, request, requestContext, config), config);
+      requestTask = createBatchableTask(name, request, requestContext, config);
+    } else {
+      requestTask = Task.async(name, () -> sendRequest(request, requestContext));
+    }
+    if (timeout.getValue() == null || timeout.getValue() <= 0) {
+      // no timeout configured
+      return requestTask;
     } else {
-      return withTimeout(Task.async(name, () -> sendRequest(request, requestContext)), config);
+      if (!_d2RequestTimeoutEnabled) {
+        return withTimeout(requestTask, timeout);
+      } else {
+        assert reconciledTimeout > 0;
```

Comment:
I think it is possible that reconciledTimeout = 0 if timeout in request context was 0?